### ISSUE TITLE
Encoding Problem

### ISSUE THREAD

**Issue by Sergios21** [tonyofrancis/Fetch#110]
I'm currently using your library at my project for downloading a large number of small files. These files are parts of a web app that needs to be downloaded locally and is displayed via a webview. Until RC5 everything was fine, I tried using RC8 and the downloaded files has a strange encoding thus the browser was not able to display them. BTW great work, keep it up.

**Comment by tonyofrancis** [comment]
@Sergios21 what kind of weird encoding are you seeing? Can you share a sample? Does this happen in another Android Studio project?

**Comment by tonyofrancis** [comment]
@Sergios21 is this still an issue?

**Comment by Sergios21** [comment]
I will be working on it tomorrow so I will let you know as soon as I am able to reproduce it. It is happening in my own Android studio project.

**Comment by Sergios21** [comment]
I've attached some files that were downloaded with the RC8 versio. As you will see they are corrupted or their encoding is wrong.

[sample.zip](https://github.com/tonyofrancis/Fetch/files/1723992/sample.zip)

**Comment by tonyofrancis** [comment]
@Sergios21 I am going to look into this some more

**Comment by tonyofrancis** [comment]
@Sergios21 Tested several urls and file downloads and they were all successful with no encoding issues. I also checked the md5 checksum of the downloaded files to make sure they matched. Could It be the device you are using? Can you test on another device?

**Comment by Sergios21** [comment]
HI Tony, I have tested on the Galaxy Note 5, on the Galaxy S8, on the Nexus 5x, on the Redmi 4x, on the Redmi note 3, One plus X, HTC U11 Plus and a Lenovo tab. We were using Amazon's S3 for storage. We have just switched to Amazon's cloudfront (I haven't tested it on cloudfront).

Following is a public S3 URL that you can test:
https://s3.amazonaws.com/myownbucket-8675309/my-internet-chart/my-webpage.html

**Comment by tonyofrancis** [comment]
@Sergios21 Just tested and I got no encoding issue. Here is the sample file downloaded with Fetch2 on a Motorola device.
[fetchwebsite.html.zip](https://github.com/tonyofrancis/Fetch/files/1777869/fetchwebsite.html.zip)

**Comment by Sergios21** [comment]
I'll be working on the project tomorrow so I will keep testing to try and get to the bottom of this.

**Comment by Sergios21** [comment]
Hello Tony, I've been trying to find the issue. I''ve narrowed down that the issue occurs from RC6 onwards RC5 works just fine.
The way I'm using Fetch is the following:
I need to download two groups of files that are later fed into a webview for offline access.
The first group are files that contain web code and images. This is essential. for the web applications to run.

These web applications have a set of multimedia assets. If these assets are not present offline, the application loads an online copy of the file. Otherwise the file is loaded from storage.

Following is my object initialization code:
`public Fetch contentFetch;
    public Fetch applicationFetch;
OkHttpClient applicationOkHttpClient = new OkHttpClient.Builder()
                .build();

        OkHttpClient contentOkHttpClient = new OkHttpClient.Builder()
                .build();

applicationFetch = new Fetch.Builder(this, "application")
                    .setDownloader(new OkHttpDownloader(applicationOkHttpClient))
                    .setDownloadConcurrentLimit(10) // Allows Fetch to download 4 downloads in Parallel.
                    .enableLogging(false)
                    .build();

contentFetch = new Fetch.Builder(this, "content")
                    .setDownloader(new OkHttpDownloader(contentOkHttpClient))
                    .setDownloadConcurrentLimit(4) // Allows Fetch to download 4 downloads in Parallel.
                    .enableLogging(false)
                    .build();
`
When using Fetch above version RC5, the applocationFetch downloads the files in the corrupt manner that I had attached. The contentFetch downloads everything normally.

All files, both content and application, are hosted on cloudfront.

Following is the method that I use to provide the local path to where each file will be saved:

`public String getSaveDir() {
        return activity.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS).toString() + "/appname/" + user.getServerUrl().replace("/", "") + "/";
    }`

**Comment by tonyofrancis** [comment]
@Sergios21 Here are a few troubleshooting suggestions.

1. Can you change the concurrent download limit to 1 for the applicationFetch and change the namespace from `application` to something else. Then try downloading  the files again

2. use the httpurlconnection downloader for the the applicationFetch and see if anything changes.

Looking forward to hearing the results.

**Comment by Sergios21** [comment]
Hello Tony,

I changed the namespace of my fetch instances and had the same results.

What actually worked was not using the OkHttpDownloader but reversing back to httpurlconnection . 

I keep using the OkHttpDownloader for my content download and it works just fine.

**Comment by tonyofrancis** [comment]
Yup I figured okHttp was encoding your content wrong. What you may want to do is check the different okhttp options and configure its settings. That may help. I will go ahead and close this issue. Open a new ticket if you have any issues.

**Comment by Sergios21** [comment]
OK Tony, thank you for your time and an awesome library. Keep up the good work.

**Comment by Sergios21** [comment]
FYI, it was a problem with the okHttp Response. My backend guys gzipped all files before uploading. Cloudfront does this on it's own. So the library read the "Accept-Encoding", "gzip, deflate, br" header and unzipped the files once but didn't know that the contents were zipped as well. Hence the encoding error. We have stopped zipping the files on our server and everything works perfectly.
